web/src/people.ts

```diff
@@ -1945,6 +1966,135 @@ export function populate_valid_user_ids(params: StateData["user_groups"]): void
+
+    const {promise_for_user_ids_in_transit, user_ids_pending_fetch} =
+        get_combined_promise_for_user_ids(user_ids_to_fetch);
+
+    // This promise will be resolved when all users are fetched.
+    fetch_users_storage.promise_for_in_transit.set(user_ids_to_fetch, {
+        promise: fetch_users_storage.promise_for_pending!,
+        resolver: fetch_users_storage.promise_resolver_for_pending!,
+    });
+
+    // Store a promise to be resolved when user_ids_pending_fetch is fetched.
+    // This avoids future requests for subset of `user_ids_pending_fetch` to wait
+    // for the completion of `user_ids_to_fetch`.
+    let resolver_for_promise_for_pending_fetch: () => void = () => {
+        // This will reassigned instantly below but Typescript thinks
+        // this function is unassigend.
+    };
+    const promise_for_pending_fetch = new Promise<void>((resolve) => {
+        resolver_for_promise_for_pending_fetch = resolve;
+    });
+    fetch_users_storage.promise_for_in_transit.set(user_ids_pending_fetch, {
+        promise: promise_for_pending_fetch,
+        resolver: resolver_for_promise_for_pending_fetch,
+    });
+
+    fetch_users_storage.promise_for_pending = undefined;
+    fetch_users_storage.promise_resolver_for_pending = undefined;
+
+    void fetch_users(user_ids_pending_fetch).then(async (fetched_users) => {
+        for (const user of fetched_users) {
+            if (user.is_active) {
+                add_active_user(user);
+            } else {
+                non_active_user_dict.set(user.user_id, user);
+                _add_user(user);
+            }
+        }
+
+        // Resolve promises waiting on this fetch after updating the data locally.
+        fetch_users_storage.promise_for_in_transit.get(user_ids_pending_fetch)!.resolver();
+        // Clean up in transit promise for this fetch.
+        fetch_users_storage.promise_for_in_transit.delete(user_ids_pending_fetch);
+        // Remove fetched users from in transit user ids.
+        fetch_users_storage.in_transit_user_ids =
+            fetch_users_storage.in_transit_user_ids.difference(user_ids_pending_fetch);
+
+        await promise_for_user_ids_in_transit;
+        // Resolve promises waiting on the complete fetch.
+        fetch_users_storage.promise_for_in_transit.get(user_ids_to_fetch)!.resolver();
+    });
+}
+
+export async function fetch_users_from_ids_internal(user_ids: number[]): Promise<unknown> {
+    // NOTE: NEVER USE THIS FUNCTION DIRECTLY.
+    // Call get_or_fetch_users_from_ids instead.
+    const unknown_user_ids = new Set(
+        user_ids.filter((user_id) => !people_by_user_id_dict.has(user_id)),
+    );
+
+    // We already have data for all requested users.
+    if (unknown_user_ids.size === 0) {
+        return undefined;
+    }
+
+    // If we have a fetch in progress which contains the unknown user_ids,
+    // return a promise that resolves when that fetch completes.
```

Comment:
This comment is wrong; it should say "If the fetches in progress contain all of the unknown user IDs, return ..."